File: Fourth Semester - Design and Programming Environments/Java/ProiectJavaNetworking/TravelPersistence/src/main/java/org/example/persistence/impl/ReservationRepositoryImpl.java
package org.example.persistence.impl;



import org.example.model.Reservation;
import org.example.model.Tour;
import org.example.model.validator.Validator;
import org.example.persistence.interfaces.ReservationRepository;
import org.example.persistence.utils.JdbcUtils;
import org.example.persistence.utils.RepositoryException;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Properties;

public class ReservationRepositoryImpl implements ReservationRepository {
    protected JdbcUtils jdbcUtils;
    protected Validator<Reservation> validator;

    public ReservationRepositoryImpl(Properties properties, Validator<Reservation> validator) {
        this.jdbcUtils = new JdbcUtils(properties);
        this.validator = validator;
    }

    private Reservation getReservationFromSet(ResultSet result) throws SQLException {
        Integer id = result.getInt("id");
        Integer numberOfReservedSeats = result.getInt("numberOfReservedSeats");

        String nameClient = result.getString("clientName");
        String contactClient = result.getString("clientContact");

        Integer tourId = result.getInt("tourId");
        String destination = result.getString("destination");
        String transportCompany = result.getString("transportCompany");
        String departure = result.getString("departureDate");
        Double price = result.getDouble("price");
        Integer numberOfAvailableSeats = result.getInt("numberOfAvailableSeats");
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
        LocalDateTime departureDate = LocalDateTime.parse(departure, formatter);
        Tour tour = new Tour(destination, transportCompany, departureDate, price, numberOfAvailableSeats);
        tour.setId(tourId);

        Reservation reservation = new Reservation(tour, nameClient, contactClient, numberOfReservedSeats);
        reservation.setId(id);
        return reservation;
    }

    @Override
    public Optional<Reservation> findOne(Integer id) {
        Connection connection = jdbcUtils.getConnection();

        String sql = "select " +
                "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                "r.clientName AS clientName, r.clientContact AS clientContact, " +
                "t.id as tourId, t.destination as destination, t.transportCompany as transportCompany, " +
                "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                "from reservation r " +
                "inner join tour t on t.id = r.tourId " +
                "where r.id = ?";
        Reservation reservation = null;
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setInt(1, id);
            try (ResultSet result = preparedStatement.executeQuery()) {
                if (result.next()) {
                    reservation = getReservationFromSet(result);
                }
            }
        } catch (SQLException exception) {
            throw new RepositoryException("Error getting connection", exception);
        }

        return Optional.ofNullable(reservation);
    }

    @Override
    public Iterable<Reservation> findAll() {
        Connection connection = jdbcUtils.getConnection();
        List<Reservation> reservations = new ArrayList<>();

        String sql = "select " +
                "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                "r.clientName AS clientName, r.clientContact AS clientContact, " +
                "t.id as tourId, t.destination as destination, t.transportCompany as transportCompany, " +
                "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                "from reservation r " +
                "inner join tour t on t.id = r.tourId ";
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            try (ResultSet result = preparedStatement.executeQuery()) {
                while (result.next()) {
                    Reservation reservation = getReservationFromSet(result);
                    if (reservation != null) {
                        reservations.add(reservation);
                    }
                }
            }
        } catch (SQLException exception) {
            throw new RepositoryException("Error getting connection", exception);
        }

        return reservations;
    }

    @Override
    public Optional<Reservation> save(Reservation entity) {
        validator.validate(entity);

        Connection connection = jdbcUtils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement("insert into reservation (tourId, clientName, clientContact, numberOfReservedSeats) values (?, ?, ?, ?)")) {
            preparedStatement.setInt(1, entity.getTour().getId());
            preparedStatement.setString(2, entity.getClientName());
            preparedStatement.setString(3, entity.getClientContact());
            preparedStatement.setInt(4, entity.getNumberOfReservedSeats());
            int result = preparedStatement.executeUpdate();
            Optional<Reservation> reservationF = findReservation(entity.getTour().getId(), entity.getClientName(), entity.getClientContact(), entity.getNumberOfReservedSeats());
            entity.setId(reservationF.get().getId());
        } catch (SQLException exception) {
            throw new RepositoryException("Error saving reservation in database", exception);
        }

        return Optional.of(entity);
    }

    @Override
    public Optional<Reservation> delete(Integer integer) {
        if (integer == null || integer < 0) {
            throw new IllegalArgumentException("Reservation id is null or negative");
        }
        Optional<Reservation> reservation = findOne(integer);
        if (reservation.isEmpty()) {
            throw new RepositoryException("This reservation doesn't exists");
        }

        Connection connection = jdbcUtils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement("delete from reservation where id = ?")) {
            preparedStatement.setInt(1, integer);

            int result = preparedStatement.executeUpdate();
        } catch (SQLException exception) {
            throw new RepositoryException("Error deleting reservation in database", exception);
        }

        return reservation;
    }

    @Override
    public Optional<Reservation> update(Reservation reservation) {
        validator.validate(reservation);

        Connection connection = jdbcUtils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement("update reservation set tourId = ?, " +
                "clientName = ?, clientName = ?, numberOfReservedSeats = ? where id = ?")) {
            preparedStatement.setInt(1, reservation.getTour().getId());
            preparedStatement.setString(2, reservation.getClientName());
            preparedStatement.setString(3, reservation.getClientContact());
            preparedStatement.setInt(4, reservation.getNumberOfReservedSeats());
            preparedStatement.setInt(5, reservation.getId());
            int result = preparedStatement.executeUpdate();
        } catch (SQLException exception) {
            throw new RepositoryException("Error updating reservation in database", exception);
        }

        return Optional.of(reservation);
    }

    @Override
    public Optional<Reservation> findReservation(Integer tourId, String nameClient, String contactClient, Integer numberOfReservedSeats) {
        String sql = "SELECT " +
                "r.id AS id, r.numberOfReservedSeats AS numberOfReservedSeats, " +
                "r.clientName AS clientName, r.clientContact AS clientContact, " +
                "t.id AS tourId, t.destination AS destination, t.transportCompany AS transportCompany, " +
                "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats AS numberOfAvailableSeats " +
                "FROM reservation r " +
                "INNER JOIN tour t ON t.id = r.tourId " +
                "WHERE r.tourId = ? AND r.clientName = ? AND r.clientContact = ? AND r.numberOfReservedSeats = ?";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql)) {

            preparedStatement.setInt(1, tourId);
            preparedStatement.setString(2, nameClient);
            preparedStatement.setString(3, contactClient);
            preparedStatement.setInt(4, numberOfReservedSeats);

            try (ResultSet result = preparedStatement.executeQuery()) {
                if (result.next()) {
                    Reservation reservation = getReservationFromSet(result);
                    return Optional.of(reservation);
                }
            }
        } catch (SQLException e) {
            throw new RepositoryException("Error finding reservation in database", e);
        }

        return Optional.empty();
    }

    @Override
    public Iterable<Reservation> findReservationsByTour(Integer tourId) {
        Connection connection = jdbcUtils.getConnection();
        List<Reservation> reservations = new ArrayList<>();

        String sql = "select " +
                "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                "r.clientName AS clientName, r.clientContact AS clientContact, " +
                "t.id as tourId, t.destination as destination, t.transportCompany as transportCompany, " +
                "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                "from reservation r " +
                "inner join tour t on t.id = r.tourId " +
                "where r.tourId = ?";
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setInt(1, tourId);
            try (ResultSet result = preparedStatement.executeQuery()) {
                while (result.next()) {
                    Reservation reservation = getReservationFromSet(result);
                    if (reservation != null) {
                        reservations.add(reservation);
                    }
                }
            }
        } catch (SQLException exception) {
            throw new RepositoryException("Error getting connection", exception);
        }

        return reservations;
    }

    @Override
    public Iterable<Reservation> findReservationsByClient(String clientName) {
        Connection connection = jdbcUtils.getConnection();

        List<Reservation> reservations = new ArrayList<>();
        String sql = "select " +
                "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                "r.clientName AS clientName, r.clientContact AS clientContact, " +
                "t.id as tourId, t.destination as destination, t.transportCompany as transportCompany, " +
                "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                "from reservation r " +
                "inner join tour t on t.id = r.tourId " +
                "where r.clientName = ?";
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setString(1, clientName);
            try (ResultSet result = preparedStatement.executeQuery()) {
                while (result.next()) {
                    Reservation reservation = getReservationFromSet(result);
                    reservations.add(reservation);
                }
            }
        } catch (SQLException exception) {
            throw new RepositoryException("Error getting connection", exception);
        }

        return reservations;
    }

}

File: Fourth Semester - Design and Programming Environments/C#/ProiectC#Networking/TravelPersistence/persistence/impl/ReservationRepositoryImpl.cs
using System.Data;
using log4net;

using TravelModel.model;
using TravelModel.model.validator;
using TravelPersistence.persistence.interfaces;
using TravelPersistence.persistence.utils;

namespace TravelPersistence.persistence.impl;

public class ReservationRepositoryImpl : IReservationRepository
{
    private static readonly ILog log =
        LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

    private IValidator<Reservation>? _reservationValidator;

    private IDictionary<String, String> properties;

    public ReservationRepositoryImpl(IDictionary<String, String> properties, IValidator<Reservation>? reservationValidator)
    {
        log.Info("Creating ReservationDatabaseRepository");

        this.properties = properties;
        this._reservationValidator = reservationValidator;
    }
    
    private Reservation GetReservationFromResult(IDataReader dataReader)
    {
        int id = Convert.ToInt32(dataReader["id"]);
        int numberOfReservedSeats = Convert.ToInt32(dataReader["numberOfReservedSeats"]);
        
        int tourId = Convert.ToInt32(dataReader["tourId"]);
        string destination = dataReader["destination"].ToString();
        string transportCompany = dataReader["transportCompany"].ToString();
        string departureDate = dataReader["departureDate"].ToString();
        double price = Convert.ToDouble(dataReader["price"]);
        int numberOfAvailableSeats = Convert.ToInt32(dataReader["numberOfAvailableSeats"]);
        DateTime departure = DateTime.ParseExact(departureDate, "yyyy-MM-dd HH:mm", System.Globalization.CultureInfo.InvariantCulture);
        Tour? tour = new Tour(destination, transportCompany, departure, price, numberOfAvailableSeats);
        tour.Id = tourId;
        
        string clientName = dataReader["clientName"].ToString();
        string clientContact = dataReader["clientContact"].ToString();
        
        Reservation? reservation = new Reservation(tour, clientName, clientContact, numberOfReservedSeats);
        reservation.Id = id;
        return reservation;
    }
    
    
    public Reservation? FindOne(int id)
    {
        log.Info("Finding Reservation");

        IDbConnection connection = DBUtils.getConnection(properties);

        if (id == null)
        {
            throw new RepositoryException("Reservation Id is null");
        }

        Reservation? reservation = null;
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select " +
                                  "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                                  "r.clientName AS clientName, r.clientContact AS clientContact, " +
                                  "t.id as tourId, t.destination as destination, t.transportCompany as transportCompany, " +
                                  "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                                  "from reservation r " +
                                  "inner join tour t on t.id = r.tourId " +
                                  "where r.id = @id";
            IDbDataParameter parameter = command.CreateParameter();
            parameter.ParameterName = "@id";
            parameter.Value = id;

            command.Parameters.Add(parameter);

            using (var dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    reservation = GetReservationFromResult(dataReader);
                    log.InfoFormat("Exiting FindOne Reservation with value: " + reservation);
                    return reservation;
                }
            }
        }

        log.InfoFormat("Exiting FindOne Reservation with value: " + reservation);
        return reservation;
    }

    public IEnumerable<Reservation> FindAll()
    {
        log.Info("Finding All Reservations");
        IDbConnection connection = DBUtils.getConnection(properties);
        IList<Reservation> reservations = new List<Reservation>();

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select " +
                                  "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                                  "r.clientName AS clientName, r.clientContact AS clientContact, " +
                                  "t.id as tourId, t.destination as destination, t.transportCompany as transportCompany, " +
                                  "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                                  "from reservation r " +
                                  "inner join tour t on t.id = r.tourId ";

            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    Reservation reservation = GetReservationFromResult(dataReader);
                    reservations.Add(reservation);
                }
            }
        }

        log.InfoFormat("Exiting FindAll reservations with value: " + reservations);
        return reservations;
    }
    
    public Reservation? FindReservation(int tourId, string clientName, string clientContact, int numberOfReservedSeats)
    {
        log.Info("Finding Reservation by all");
        IDbConnection connection = DBUtils.getConnection(properties);

        if (tourId == null || clientName == null || clientContact == null || numberOfReservedSeats == null || numberOfReservedSeats == 0)
        {
            throw new RepositoryException("Reservation tourId or clientName or clientContact or numberOfReservedSeats are null");
        }
        
        Reservation? reservation = null;
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select " +
                                  "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                                  "r.clientName AS clientName, r.clientContact AS clientContact, " +
                                  "t.id as tourId, t.destination as destination, t.transportCompany as transportCompany, " +
                                  "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                                  "from reservation r " +
                                  "inner join tour t on t.id = r.tourId " +
                                  "where r.tourId = @tourId and r.clientName = @clientName and r.clientContact = @clientContact and r.numberOfReservedSeats = @numberOfReservedSeats";
            IDbDataParameter tourParameter = command.CreateParameter();
            tourParameter.ParameterName = "@tourId";
            tourParameter.Value = tourId;
            command.Parameters.Add(tourParameter);
            
            IDbDataParameter clientNameParameter = command.CreateParameter();
            clientNameParameter.ParameterName = "@clientName";
            clientNameParameter.Value = clientName;
            command.Parameters.Add(clientNameParameter);
            
            IDbDataParameter clientContactParameter = command.CreateParameter();
            clientContactParameter.ParameterName = "@clientContact";
            clientContactParameter.Value = clientContact;
            command.Parameters.Add(clientContactParameter);
            
            IDbDataParameter numberOfReservedSeatsParameter = command.CreateParameter();
            numberOfReservedSeatsParameter.ParameterName = "@numberOfReservedSeats";
            numberOfReservedSeatsParameter.Value = numberOfReservedSeats;
            command.Parameters.Add(numberOfReservedSeatsParameter);

            using (var dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    reservation = GetReservationFromResult(dataReader);
                    log.InfoFormat("Exiting Find Reservation with value: " + reservation);
                    return reservation;
                }
            }
        }
        log.InfoFormat("Exiting Find Reservation  with value: " + reservation);
        return reservation;
    }

    
    public Reservation? Save(Reservation reservation)
    {
        log.Info("Saving Reservation");
        IDbConnection connection = DBUtils.getConnection(properties);

        _reservationValidator.Validate(reservation);

        Reservation? reservationFound = FindReservation(reservation.Tour.Id, reservation.ClientName, reservation.ClientContact, reservation.NumberOfReservedSeats);
        if (reservationFound != null)
        {
            log.InfoFormat("Reservation already exists with value: " + reservationFound);
            throw new RepositoryException("Reservation already exists");
        }

        using (var command = connection.CreateCommand())
        {
            command.CommandText =
                "insert into reservation (tourId, clientName, clientContact, numberOfReservedSeats) values (@tourId, @clientName, @clientContact, @numberOfReservedSeats)";
            IDbDataParameter tourParameter = command.CreateParameter();
            tourParameter.ParameterName = "@tourId";
            tourParameter.Value = reservation.Tour.Id;
            command.Parameters.Add(tourParameter);
            
            IDbDataParameter clientNameParameter = command.CreateParameter();
            clientNameParameter.ParameterName = "@clientName";
            clientNameParameter.Value = reservation.ClientName;
            command.Parameters.Add(clientNameParameter);
            
            IDbDataParameter clientContactParameter = command.CreateParameter();
            clientContactParameter.ParameterName = "@clientContact";
            clientContactParameter.Value = reservation.ClientContact;
            command.Parameters.Add(clientContactParameter);
            
            IDbDataParameter numberOfReservedSeatsParameter = command.CreateParameter();
            numberOfReservedSeatsParameter.ParameterName = "@numberOfReservedSeats";
            numberOfReservedSeatsParameter.Value = reservation.NumberOfReservedSeats;
            command.Parameters.Add(numberOfReservedSeatsParameter);

            command.ExecuteNonQuery();

            Reservation? reservationF = FindReservation(reservation.Tour.Id, reservation.ClientName, reservation.ClientContact, reservation.NumberOfReservedSeats);
            reservation.Id = reservationF.Id;
            log.InfoFormat("Reservation saved with value: " + reservation);
        }

        return reservation;
    }

    public Reservation? Delete(int id)
    {
        log.Info("Deleting Reservation");
        IDbConnection connection = DBUtils.getConnection(properties);

        if (id == null)
        {
            throw new RepositoryException("Reservation Id is null");
        }
        
        Reservation? reservationFound = FindOne(id);
        if (reservationFound == null)
        {
            log.InfoFormat("Reservation is not found");
            throw new RepositoryException("Reservation is not found");
        }

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "delete from reservation where id = @id";
            IDbDataParameter parameter = command.CreateParameter();
            parameter.ParameterName = "@id";
            parameter.Value = id;
            command.Parameters.Add(parameter);
        
            command.ExecuteNonQuery();

            log.InfoFormat("Reservation deleted with value: " + reservationFound);
        }

        return reservationFound;
    }

    public Reservation? Update(Reservation reservation)
    {
        log.Info("Updating Reservation");
        IDbConnection connection = DBUtils.getConnection(properties);

        _reservationValidator.Validate(reservation);

        Reservation? reservationFound = FindOne(reservation.Id);
        if (reservationFound == null)
        { 
            log.InfoFormat("Reservation is not found");
            throw new RepositoryException("Reservation is not found");
        }

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "update reservation set tourId = @tourId, numberOfReservedSeats = @numberOfReservedSeats where id = @id";
            IDbDataParameter tourParameter = command.CreateParameter();
            tourParameter.ParameterName = "@tourId";
            tourParameter.Value = reservation.Tour.Id;
            command.Parameters.Add(tourParameter);
            
            IDbDataParameter numberOfReservedSeatsParameter = command.CreateParameter();
            numberOfReservedSeatsParameter.ParameterName = "@numberOfReservedSeats";
            numberOfReservedSeatsParameter.Value = reservation.NumberOfReservedSeats;
            command.Parameters.Add(numberOfReservedSeatsParameter);
            
            IDbDataParameter idParameter = command.CreateParameter();
            idParameter.ParameterName = "@id";
            idParameter.Value = reservation.Id;
            command.Parameters.Add(idParameter);
            
            command.ExecuteNonQuery();
            
            log.InfoFormat("Agency added with value: " + reservation);
        }

        return reservation;
    }

    public IEnumerable<Reservation> FindReservationsByTour(int tourId)
    {
        log.Info("Finding Reservations By Tour");
        IDbConnection connection = DBUtils.getConnection(properties);
        IList<Reservation> reservations = new List<Reservation>();

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select " +
                                  "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                                  "r.clientName AS clientName, r.clientContact AS clientContact, " +
                                  "t.id as tourId, t.destination as destination, t.transportCompany as transportCompany, " +
                                  "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                                  "from reservation r " +
                                  "inner join tour t on t.id = r.tourId " +
                                  "where r.tourId = @tourId";
            IDbDataParameter tourIdParameter = command.CreateParameter();
            tourIdParameter.ParameterName = "@tourId";
            tourIdParameter.Value = tourId;
            command.Parameters.Add(tourIdParameter);
            
            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    Reservation reservation = GetReservationFromResult(dataReader);
                    reservations.Add(reservation);
                }
            }
        }

        log.InfoFormat("Exiting Find reservations by tour with value: " + reservations);
        return reservations;
    }
    
    public IEnumerable<Reservation> FindReservationsByClient(string clientName)
    {
        log.Info("Finding Reservations By Tour");
        IDbConnection connection = DBUtils.getConnection(properties);
        IList<Reservation> reservations = new List<Reservation>();

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select " +
                                  "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                                  "r.clientName AS clientName, r.clientContact AS clientContact, " +
                                  "t.id as tourId, t.destination as destination, t.transportCompany as transportCompany, " +
                                  "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                                  "from reservation r " +
                                  "inner join tour t on t.id = r.tourId " +
                                  "where r.clientName = @clientName";
            IDbDataParameter clientNameParameter = command.CreateParameter();
            clientNameParameter.ParameterName = "@clientName";
            clientNameParameter.Value = clientName;
            command.Parameters.Add(clientNameParameter);
            
            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    Reservation reservation = GetReservationFromResult(dataReader);
                    reservations.Add(reservation);
                }
            }
        }

        log.InfoFormat("Exiting Find reservations by client with value: " + reservations);
        return reservations;
    }
}